### PR TITLE
Call none

### PR DESCRIPTION
**Pull request by kaspernielsen96**
**Context:**
Adding a method where one can call the ansatz with `None` for some arguments. This will return a an updated ansatz with on fewer variables (wires).
**Description of the Change:**
Adding a method to the `DiffOpPolyExpAnsatz` where one can evaluate only some of the wires, and thus returning a new updated ansatz.
**Benefits:**
Allows for reduction in the size of the ansatz. Can also be used with `BtoQ` to do a Homodyne measurement.

### CODE REVIEW

File: mrmustard/physics/ansatze.py [hunk 1]
```diff
@@ -787,6 +787,84 @@ def __call__(self, z: Batch[Vector]) -> Scalar:
             )  # (b_arg)
         return val
 
+    def call_none(self, z: Batch[Vector]) -> DiffOpPolyExpAnsatz:
+        r"""
+        Updates the ansatz from calling the ansatz on some wires, while leaving the rest as None.
```
Comment (ziofil) [inline mrmustard/physics/ansatze.py:792]:
```suggestion
        Returns a new ansatz that corresponds to currying (partially evaluate) the current one.
        For example, if ``self`` represents the function ``F(z1,z2)``, the call ``self.call_none([np.array([1.0, None]])``
        returns ``F(1.0, z2)`` as a new ansatz with a single variable.
```

File: mrmustard/physics/ansatze.py [hunk 2]
```diff
@@ -787,6 +787,84 @@ def __call__(self, z: Batch[Vector]) -> Scalar:
...
+        Note that the batch of the triple and argument in this method is handled parwise, unlike the regular call where the batch over the triple is a superposition.
```
Comment (ziofil) [inline mrmustard/physics/ansatze.py:793]: what do you mean? can you expand a bit?

File: mrmustard/physics/ansatze.py [hunk 3]
```diff
@@ -787,6 +787,84 @@ def __call__(self, z: Batch[Vector]) -> Scalar:
...
+        Args:
+            z: slice in C^n where the function is evaluated, while unevaluated along other axes of the space.
+
+        Returns:
+            A new ansatz, which is a "slice" of the old one.
+        """
+
+        def call_none_single(Ai, bi, ci, zi):
```
Comment (ziofil) [inline mrmustard/physics/ansatze.py:801]: Can we move this outside of the scope of ``call_none``? You could add it as a private method of this class with an underscore in front.

File: mrmustard/physics/ansatze.py (hunk 2 above)
Comment (kaspernielsen96) [reply mrmustard/physics/ansatze.py]: I added a bit more, let me know if it is still not clear.

File: mrmustard/physics/ansatze.py [hunk 4]
```diff
@@ -727,7 +727,19 @@ def from_function(cls, fn: Callable, **kwargs: Any) -> DiffOpPolyExpAnsatz:
         ret._kwargs = kwargs
         return ret
 
-    def __call__(self, z: Batch[Vector]) -> Scalar:
+    def __call__(self, z: Batch[Vector]) -> Union[Scalar, DiffOpPolyExpAnsatz]:
+        r"""
+        Returns either the value of the ansatz or a new ansatz depending on the argument.
+        If the argument contains None, returns a new ansatz.
+        If the argument only contains numbers, returns the value of the ansatz at that argument.
+        Note that the batch dimensions are handled differently in the two cases. See subfunctions for furhter information.
```
Comment (ziofil) [inline mrmustard/physics/ansatze.py:735]: please add the "Args:" and "Returns:" section of the docstring

File: mrmustard/physics/ansatze.py [hunk 5]
```diff
@@ -727,7 +727,19 @@ def from_function(cls, fn: Callable, **kwargs: Any) -> DiffOpPolyExpAnsatz:
...
+        """
+        if (np.array(z) == None).any():
+            return self._call_none(z)
+        else:
+            return self._call_all(z)
+
+    def _call_all(self, z: Batch[Vector]) -> DiffOpPolyExpAnsatz:
         r"""
         Value of this ansatz at ``z``. If ``z`` is batched a value of the function at each of the batches are returned.
         If ``Abc`` is batched it is thought of as a linear combination, and thus the results are added linearly together.
```
Comment (ziofil) [inline mrmustard/physics/ansatze.py:745]: please add the "Args:" and "Returns:" section of the docstring

File: mrmustard/physics/ansatze.py [hunk 6]
```diff
@@ -794,6 +806,111 @@ def __call__(self, z: Batch[Vector]) -> Scalar:
             )  # (b_arg)
         return val
 
+    def _call_none_single(self, Ai, bi, ci, zi):
+        r"""
+        Helper function for the call_none method. Returns the new triple.
```
Comment (ziofil) [inline mrmustard/physics/ansatze.py:811]: please add the "Args:" and "Returns:" section of the docstring

File: mrmustard/physics/ansatze.py [hunk 7]
```diff
@@ -794,6 +806,111 @@ def __call__(self, z: Batch[Vector]) -> Scalar:
...
+
+        Args:
+            z: slice in C^n where the function is evaluated, while unevaluated along other axes of the space.
+
+        Returns:
+            A new ansatz.
+        """
+
+        batch_abc = self.batch_size
+        print(batch_abc)
```
Comment (ziofil) [inline mrmustard/physics/ansatze.py:894]: remove?

File: mrmustard/physics/ansatze.py [hunk 8]
```diff
@@ -794,6 +806,111 @@ def __call__(self, z: Batch[Vector]) -> Scalar:
...
+        Abc = []
+        if batch_abc == 1 and batch_arg > 1:
+            for i in range(batch_arg):
+                Abc.append(self._call_none_single(self.A[0], self.b[0], self.c[0], z[i]))
+        elif batch_arg == 1 and batch_abc > 1:
+            for i in range(batch_abc):
+                Abc.append(self._call_none_single(self.A[i], self.b[i], self.c[i], z[0]))
+        elif batch_abc == batch_arg:
+            for i in range(batch_abc):
+                print(i)
```
Comment (ziofil) [inline mrmustard/physics/ansatze.py:905]: remove?

**Review by ziofil — APPROVED** [review]
👍

**Pull request merged**